src/parser.js

```diff
@@ -50,49 +44,90 @@ const ignore = async (stream) => {
   }
 }
 
-async function * parser (stream, options) {
-  for await (const part of multipart(stream, options.boundary)) {
-    const partHeader = parseHeader(part.headers)
+async function * parseEntry (stream, options) {
+  let entry = {}
 
-    if (isDirectory(partHeader.mime)) {
-      yield {
-        type: 'directory',
-        name: partHeader.name
+  for await (const part of stream) {
+    let type
+
+    if (part.headers['content-type']) {
+      type = Content.type(part.headers['content-type'])
+
+      if (type.boundary) {
+        // recursively parse nested multiparts
+        yield * parser(part.body, {
+          ...options,
+          boundary: type.boundary
+        })
+
+        continue
       }
+    }
+
+    if (!part.headers['content-disposition']) {
+      throw new Error('No content disposition in multipart part')
+    }
 
-      await ignore(part.body)
+    const disposition = parseDisposition(part.headers['content-disposition'])
 
-      continue
+    if (disposition.name.includes('mtime')) {
```

Comment:
Ok so, metadata must preceed the file data if set. Can you document the proposal so we can check this is ok with the go-ipfs team?

Reply:
This has been changed to be passed as a header instead of a multipart part